rfcs/20250314-graph-api-bottom-right-causal-mask/README.md

```diff
@@ -0,0 +1,382 @@
+                            Pointwise_attributes()
+                                .set_name("row_idx_add_skv")
+                                .set_mode(PointwiseMode_t::ADD)
+                                .set_compute_data_type(DataType_t::INT32));
+```
+
+To support similar usage in oneDNN Graph, extend the `logical_tensor` API as follows:
+
+```cpp
+/// Constructs a logical tensor object with scalar value.
+///
+/// @param tid Logical tensor ID.
+/// @param dtype Elements data type.
+/// @param scalar Scalar value.
+logical_tensor(size_t tid, data_type dtype, void *scalar) {
+    dnnl_graph_logical_tensor_t val;
+    error::wrap_c_api(dnnl_graph_logical_tensor_init_with_scalar(
+                              &val, tid, convert_to_c(dtype), scalar),
+            "could not create logical_tensor with scalar");
+    data = val;
+}
+```
+
+And update the `dnnl_graph_logical_tensor_t` structure to support saving of scalars.
+
+```cpp
+typedef struct {
+    size_t id;
+    int ndims;
+    dnnl_dims_t dims;
+    dnnl_data_type_t data_type;
+    dnnl_graph_tensor_property_t property;
+    dnnl_graph_layout_type_t layout_type;
+    union {
+        dnnl_dims_t strides;
+        size_t layout_id;
+    } layout;
+
+    // New fields for scalar logical tensor support:
+    bool pass_by_scalar;
+    union {
+        float f32_value;
+        int32_t s32_value;
+    } scalar;
+} dnnl_graph_logical_tensor_t;
+```
+
+#### Pros
+
+- No need to create GPU tensors for s_kv and s_q scalars,
+  and other scalars like attn_scale, -inf.
+- Consistent subgraph design with the top-left implicit causal mask.
+- Align with cuDNN Graph's usage.
+
+#### Cons
+
+- Requires extending the Graph API to support scalar logical tensors
+  (an ABI-breaking change that must wait for the oneDNN v4.0 release).
+
+### Option 1+2: Pass s_kv and s_q as Scalar Tensors While Indicating Scalar Property in Logical Tensors
+
+To mitigate the ABI-breaking change introduced in Option 2, this
+option suggests passing s_kv and s_q as scalar tensors while
+indicating their scalar nature at the logical tensor level.
+
+The current `dnnl_graph_logical_tensor_t` structure:
+
+```cpp
+typedef struct {
+    size_t id;
+    int ndims;
+    dnnl_dims_t dims;
+    dnnl_data_type_t data_type;
+    dnnl_graph_tensor_property_t property;
+    dnnl_graph_layout_type_t layout_type;
+    union {
+        dnnl_dims_t strides;
+        size_t layout_id;
+    } layout;
+} dnnl_graph_logical_tensor_t;
+```
+
+Currently supported dnnl_graph_tensor_property_t values:
+- `dnnl_graph_tensor_property_undef`: Undefined tensor property.
+- `dnnl_graph_tensor_property_variable`: The tensor may change during computation
+  or between iterations.
+- `dnnl_graph_tensor_property_constant`: The tensor remains unchanged during computation and across
+  iterations, allowing optimization and caching (e.g., constant weight tensors in inference scenarios).
```

Comment:
I think this part can be safely removed. But I suggest to add more information and usage about the newly added `host_scalar` property here. I think we had quite a few discussions about setting the property under different runtime combinations. Better to have them documented here.

Reply:
Thanks for the suggestion, updated.